lib/providers/ws/services/oracles.js

```diff
@@ -0,0 +1,202 @@
+/*
+ * ISC License (ISC)
+ * Copyright 2018 aeternity developers
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
+ *  REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
+ *  AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
+ *  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
+ *  LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
+ *  OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
+ *  PERFORMANCE OF THIS SOFTWARE.
+ */
+
+
+const EventEmitter = require('events').EventEmitter
+const {actions} = require('../types')
+const {AeSubscription, OracleRegistrationSubscription} = require('../subscriptions')
+
+class Oracles extends EventEmitter {
+
+  /**
+   *
+   */
+  constructor(wsProvider) {
+    super()
+
+    let _this = this
+    this._ws = wsProvider
+
+    this._ws.addSubscription(new OracleRegistrationSubscription(this._ws))
+
+    this._ws.addSubscription(
+      new AeSubscription({
+        action: actions.QUERY,
+        origin: 'oracle',
+        update: (data) => {
+          _this._ws.emit ('query', data.payload['query_id'])
+        }
+      })
+    )
+
+    this._ws.addSubscription(
+      new AeSubscription({
+        action: actions.SUBSCRIBE,
+        origin: 'oracle',
+        update: (data) => {
+          _this._ws.emit ('subscribed', data.payload['subscribed_to'])
+        }
+      })
+    )
+
+    this._ws.addSubscription(
+      new AeSubscription({
+        action: actions.NEW_ORACLE_QUERY,
+        origin: 'chain',
+        update: (data) => _this._ws.emit ('newQuery', data.payload)
+      })
+    )
+
+    this._ws.addSubscription(
+      new AeSubscription({
+        action: actions.NEW_ORACLE_RESPONSE,
+        origin: 'chain',
+        update: (data) => _this._ws.emit ('response', data.payload)
+      })
+    )
+
+  }
+
+  /**
+   * Registers an oracle on the blockchain
+   *
+   * @param queryFormat format of the input
+   * @param responseFormat format of the response
+   * @param queryFee regular costs to post a query
+   * @param ttl relative number of blocks before a query is dropped
+   * @param fee the fee to register the oracle
+   */
+  register(queryFormat, responseFormat, queryFee, ttl, fee) {
+    let data = {
+      'target': 'oracle',
+      'action': 'register',
+      'payload': {
+        'type': 'OracleRegisterTxObject',
+        'vsn': 1,
+        'account': this.account,
+        'query_format': queryFormat,
+        'response_format': responseFormat,
+        'query_fee': queryFee,
+        'ttl': {'type': 'delta', 'value': ttl},
+        'fee': fee
+      }
```

Comment:
I feel like in Javascript world, you can make the `register` and `subscribe` calls even prettier by making use of its async nature. In this case you could return a Promise that gets resolved as soon as the registration acknoledgement comes in. This would allow you to write code like this:
```
Oracles.register(...).then(() => Oracles.subscribe())
```
or the even prettier variant:
```
oracle_registration = await Oracles.register(...);
oracle_subscription = await Oracles.subscribe(...);
```

Reply:
I agree, that this would be nice from a programmatical perspectiv, but I am not sure, if I should promisify the oracles methods. The websocket implementations I have been introspecting were all event-based.